src/internal/storage/server.go

```diff
@@ -87,3 +104,189 @@ func getOrCreateKey(ctx context.Context, keyStore chunk.KeyStore, name string) (
 	res, err := keyStore.Get(ctx, name)
 	return res, errors.EnsureStack(err)
 }
+
+// TODO: Copy file.
+func (s *Server) CreateFileset(server storage.Fileset_CreateFilesetServer) error {
+	ctx := server.Context()
+	var id *fileset.ID
+	if err := s.Filesets.WithRenewer(ctx, defaultTTL, func(ctx context.Context, renewer *fileset.Renewer) error {
+		// TODO: Validator
+		opts := []fileset.UnorderedWriterOption{fileset.WithRenewal(defaultTTL, renewer)}
+		uw, err := s.Filesets.NewUnorderedWriter(ctx, opts...)
+		if err != nil {
+			return err
+		}
+		for {
+			msg, err := server.Recv()
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					break
+				}
+				return err
+			}
+			switch mod := msg.Modification.(type) {
+			case *storage.CreateFilesetRequest_AppendFile:
+				if err := uw.Put(ctx, mod.AppendFile.Path, "", true, bytes.NewReader(mod.AppendFile.Data.Value)); err != nil {
+					return err
+				}
+			case *storage.CreateFilesetRequest_DeleteFile:
+				if err := uw.Delete(ctx, mod.DeleteFile.Path, ""); err != nil {
+					return err
+				}
+			}
+		}
+		id, err = uw.Close(ctx)
+		return err
+	}); err != nil {
+		return err
+	}
+	return server.SendAndClose(&storage.CreateFilesetResponse{
+		FilesetId: id.HexString(),
+	})
+}
+
+func (s *Server) ReadFileset(request *storage.ReadFilesetRequest, server storage.Fileset_ReadFilesetServer) error {
+	ctx := server.Context()
+	id, err := fileset.ParseID(request.FilesetId)
+	if err != nil {
+		return err
+	}
+	fs, err := s.Filesets.Open(ctx, []fileset.ID{*id})
+	if err != nil {
+		return err
+	}
+	// Compute the intersection of the path range filters.
+	pathRange := &index.PathRange{}
+	for _, f := range request.Filters {
+		switch f := f.Filter.(type) {
+		case *storage.FileFilter_PathRange:
+			// Return immediately if the path ranges are disjoint.
+			if pathRange.Lower != "" && f.PathRange.Upper <= pathRange.Lower ||
+				pathRange.Upper != "" && f.PathRange.Lower >= pathRange.Upper {
+				return nil
```

Comment:
It would be worth returning a validation error in the future on the file filters to help clients catch disjoint filters more easily